sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c

```diff
@@ -0,0 +1,410 @@
+                          _nib_get_if(tmp));
+                    clist_rpush(&_next_removable, (clist_node_t *)tmp);
+                }
+                tmp = (_nib_t *)clist_lpop(&_next_removable);
+            } while (tmp != first);
+        }
+    }
+    return nib;
+}
+
+_nib_t *_nib_iter(const _nib_t *last)
+{
+    bool found = false;
+
+    for (unsigned i = 0; i < GNRC_IPV6_NIB_NUMOF; i++) {
+        _nib_t *node = &_nodes[i];
+
+        if (node->mode != _EMPTY) {
+            if ((last == NULL)) {
+                return node;
+            }
+            else if (!found && (last == node)) {
+                found = true;
+            }
+            else if (found) {
+                return node;
+            }
+        }
+    }
+    return NULL;
+}
+
+_nib_t *_nib_get(const ipv6_addr_t *addr, unsigned iface)
+{
+    assert(addr != NULL);
+    DEBUG("nib: Getting on-link node entry (addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+    for (unsigned i = 0; i < GNRC_IPV6_NIB_NUMOF; i++) {
+        _nib_t *node = &_nodes[i];
+
+        if ((node->mode != _EMPTY) &&
+            /* either requested or current interface undefined or interfaces equal */
+            ((_nib_get_if(node) == 0) || (iface == 0) ||
+             (_nib_get_if(node) == iface)) &&
+            ipv6_addr_equal(&node->ipv6, addr)) {
+            DEBUG("  Found %p\n", (void *)node);
+            return node;
+        }
+    }
+    DEBUG("  No suitable entry found\n");
+    return NULL;
+}
+
+void _nib_nc_set_reachable(_nib_t *nib)
+{
+    _nib_iface_t *iface = _nib_iface_get(_nib_get_if(nib));
+    DEBUG("nib: set %s%%%u reachable (reachable time = %u)\n",
+          ipv6_addr_to_str(addr_str, &nib->ipv6, sizeof(addr_str)),
+          _nib_get_if(nib), iface->reach_time);
+    nib->info &= ~GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK;
+    nib->info |= GNRC_IPV6_NIB_NC_INFO_NUD_STATE_REACHABLE;
+    /* TODO add event for state change to STALE to event timer*/
+    (void)iface;
+}
+
+void _nib_nc_remove(_nib_t *nib)
+{
+    DEBUG("nib: remove from neighbor cache (addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, &nib->ipv6, sizeof(addr_str)),
+          _nib_get_if(nib));
+    nib->mode &= ~(_NC);
+    /* TODO: remove NC related timers */
+    _nib_clear(nib);
+}
+
+_nib_dr_t *_nib_drl_add(const ipv6_addr_t *router_addr, unsigned iface)
+{
+    _nib_dr_t *def_router = NULL;
+
+    DEBUG("nib: Allocating default router list entry entry "
+          "(router_addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, router_addr, sizeof(addr_str)), iface);
+    for (unsigned i = 0; i < GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF; i++) {
+        _nib_dr_t *tmp = &_def_routers[i];
+        _nib_t *tmp_node = tmp->next_hop;
+
+        if ((tmp_node != NULL) &&
+            (ipv6_addr_equal(router_addr, &tmp_node->ipv6)) &&
+            (_nib_get_if(tmp_node) == iface)) {
```

Comment:
here again, checking for the face first and then for the ipv6 address might be a better way to go.